util/ubaboot.py

```diff
@@ -0,0 +1,303 @@
+            pgm.extend(self.__dev_read(cmd, addr, 0, nb))
+        return pgm
+
+    def __write_mem(self, cmd, base, pgm, blksz):
+        for offset in range(0, len(pgm), blksz):
+            block = pgm[offset:offset + blksz]
+            padding = ((len(block) + blksz - 1) & ~(blksz - 1)) - len(block)
+            block += bytearray(b'\xff') * padding
+            self.__dev_write(cmd, base + offset, 0, block)
+
+
+def main():
+    args = parse_args()
+
+    if args.check:
+        sys.exit(0)
+
+    vid, pid = args.dev
+    dev = usb.core.find(idVendor=vid, idProduct=pid)
+    if not dev:
+        if (args.gui):
+            gui = Gui(vid, pid)
+            gui.run()
+            if gui.cancelled:
+                raise ValueError('Cancelled')
+            dev = usb.core.find(idVendor=vid, idProduct=pid)
+        else:
+            if (args.wait):
+                while True:
+                    dev = usb.core.find(idVendor=vid, idProduct=pid)
+                    if dev:
+                        break
+                    print("Cannot open ubaboot device. Retrying...")
+                    time.sleep(1)
+            else:
+                raise ValueError('cannot open ubaboot device')
+
+    ubaboot = UbabootDevice(dev)
+
+    if args.mode == 'stat':
+        to_hex = lambda v: '%02X' % (v,)
+        sig = ubaboot.get_signature()
+        lofuse, lock, extfuse, hifuse = bytearray(ubaboot.get_lock())
+        fuses = [lofuse, hifuse, extfuse]
+        print('Signature:', ''.join(to_hex(v) for v in bytearray(sig)), \
+            'Fuses: L:{0} H:{1} E:{2}'.format(*(to_hex(v) for v in fuses)), \
+            'Lock: {0}'.format(to_hex(lock)))
+
+    if args.mode in ('verify', 'write'):
+        with open(args.file) as f:
+            pgm = read_ihex(f)
+        print('Loaded', len(pgm), 'bytes from', args.file)
+
+    if args.mode in ('read', 'write', 'verify'):
+        if args.type == 'flash':
+            read, write = ubaboot.read_flash, ubaboot.write_flash
+        elif args.type == 'eeprom':
+            read, write = ubaboot.read_eeprom, ubaboot.write_eeprom
+
+    if args.mode == 'read':
+        start = args.start
+        count = args.count
+        mem = read(start, count)
+        mem = mem.rstrip(bytearray(b'\xff'))
+        for offset in range(0, len(mem), 16):
+            print(IHexLine(addr=start + offset, type=0, data=mem[offset:offset + 16]))
+        print(IHexLine(addr=0, type=1, data=''))
+
+    if args.mode == 'write':
+        write(0, pgm)
+        print('Wrote', len(pgm), 'bytes OK')
+
+    if args.mode in ('verify', 'write'):
+        actual = read(0, len(pgm))
+        for pos in range(len(pgm)):
+            if actual[pos] != pgm[pos]:
+                raise ValueError('mismatch at offset 0x%x' % (pos,))
+        print('Verified', len(pgm), 'bytes OK')
+
+    if args.mode in ('write', 'reboot') and args.reboot:
+        ubaboot.reboot()
+        print('Rebooting')
+
+
+class Gui:
```

Comment:
We would want to keep the changes as minimal as possible to mitigate maintenance issues. As such, I dont think we would want to pick up this `Gui` section of the code and its additional handling above.

Reply:
It makes sense. Removed.

About the upstream, I'll sent it too.

Reply:
All done.